src/util/primitives.js

```diff
@@ -22,6 +23,49 @@ class Ray {
         vec3.copy(out, intersection);
         return true;
     }
+
+    closestPointOnSphere(center: vec3, r: number, out: vec3): boolean {
+        assert(vec3.squaredLength(this.dir) > 0.0 && r >= 0.0);
+        if (vec3.equals(this.pos, center) || r === 0.0) {
+            out = [0.0, 0.0, 0.0];
+            return false;
+        }
+
+        const centerToP = vec3.sub([], this.pos, center);
+        const a = vec3.dot(this.dir, this.dir);
+        const b = 2.0 * vec3.dot(centerToP, this.dir);
+        const c = vec3.dot(centerToP, centerToP) - r * r;
+        const d = b * b - 4 * a * c;
+
+        if (d < 0.0) {
+            // No intersection, find distance between closest points
+            vec3.scale(centerToP, centerToP, -1.0);
+
+            const t = Math.max(vec3.dot(centerToP, this.dir), 0.0);
+            const pointOnRay = vec3.add([], this.pos, vec3.scale([], this.dir, t));
+            const pointToGlobe = vec3.sub([], center, pointOnRay);
+
+            const pointToGlobeLength = vec3.length(pointToGlobe);
+            vec3.scale(pointToGlobe, pointToGlobe, 1.0 - r / pointToGlobeLength);
+            vec3.sub(out, vec3.add([], pointOnRay, pointToGlobe), center);
```

Comment:
I have a bad habit of writing allocation unfriendly javascript code especially when prototyping things. Perhaps we could do an optimization pass at some point for reducing unnecessary array allocations when invoking vecN functions?

Reply:
This code path is globe only, we can defer that and address in separate PR. I'll make sure to capture these improvements.